lib/queuePopulator/IngestionReader.js

```diff
@@ -0,0 +1,233 @@
+            });
+            return done();
+        }
+        if (logRes.info.start === null || logRes.log === null) {
+            return done(null);
+        }
+
+        this._setEntryBatch(entriesToPublish);
+
+        logRes.log.on('data', record => {
+            logStats.nbLogRecordsRead += 1;
+            record.entries.forEach(entry => {
+                logStats.nbLogEntriesRead += 1;
+                this._processLogEntry(batchState, record, entry,
+                    this.bucketdPrefix);
+            });
+        });
+        logRes.log.on('error', err => {
+            this.log.error('error fetching entries from log',
+                { method: 'LogReader._processPrepareEntries',
+                    error: err });
+            return done(err);
+        });
+        logRes.log.on('end', () => {
+            this.log.debug('ending record stream');
+            return done();
+        });
+        return undefined;
+    }
+
+    _processPublishEntries(batchState, done) {
+        const { entriesToPublish, logRes, logStats } = batchState;
+
+        if (this.remoteLogOffset /* &&
+        config.queuePopulator.logSource.indexOf('ingestion') > -1 */) {
+            batchState.nextLogOffset = this.remoteLogOffset;
+            // return done();
+        } else {
+            batchState.nextLogOffset =
+            logRes.info.start + logStats.nbLogRecordsRead;
+        }
+
+        return async.each(Object.keys(entriesToPublish), (topic, done) => {
+            const topicEntries = entriesToPublish[topic];
+            if (topicEntries.length === 0) {
+                return done();
+            }
+            return async.series([
+                done => this._setupProducer(topic, done),
+                done => this._producers[topic].send(topicEntries, done),
+            ], err => {
+                if (err) {
+                    this.log.error(
+                        'error publishing entries from log to topic',
+                        { method: 'LogReader._processPublishEntries',
+                          topic,
+                          entryCount: topicEntries.length,
+                          error: err });
+                    return done(err);
+                }
+                this.log.debug('entries published successfully to topic',
+                               { method: 'LogReader._processPublishEntries',
+                                 topic, entryCount: topicEntries.length });
+                batchState.publishedEntries[topic] = topicEntries;
+                return done();
+            });
+        }, err => {
+            // TODO: On error, produce error metrics
+            if (err) {
+                return done(err);
+            }
+            // Find the CRR Class extension
+            const crrExtension = this._extensions.find(ext => (
+                ext instanceof ReplicationQueuePopulator
+            ));
+            const extMetrics = crrExtension.getAndResetMetrics();
```

Comment:
I think we should remove metrics here, may duplicate crr metrics since publishing to same extension with same redis keys